library/include/DataFuncs.h

```diff
@@ -34,229 +34,106 @@ distribution.
-#undef FW_TARGS
 
-#define FW_TARGS class A1, class A2, class A3, class A4
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4))
-INSTANTIATE_WRAPPERS(4, (A1,A2,A3,A4), (OSTREAM_ARG,A1,A2,A3,A4),
-                        (vA1,vA2,vA3,vA4), (out,vA1,vA2,vA3,vA4),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);)
-#undef FW_TARGS
-
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5))
-INSTANTIATE_WRAPPERS(5, (A1,A2,A3,A4,A5), (OSTREAM_ARG,A1,A2,A3,A4,A5),
-                        (vA1,vA2,vA3,vA4,vA5), (out,vA1,vA2,vA3,vA4,vA5),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5);)
-#undef FW_TARGS
-
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5, class A6
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5,A6))
-INSTANTIATE_WRAPPERS(6, (A1,A2,A3,A4,A5,A6), (OSTREAM_ARG,A1,A2,A3,A4,A5,A6),
-                        (vA1,vA2,vA3,vA4,vA5,vA6), (out,vA1,vA2,vA3,vA4,vA5,vA6),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5); LOAD_ARG(A6);)
-#undef FW_TARGS
-
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5, class A6, class A7
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5,A6,A7))
-INSTANTIATE_WRAPPERS(7, (A1,A2,A3,A4,A5,A6,A7), (OSTREAM_ARG,A1,A2,A3,A4,A5,A6,A7),
-                        (vA1,vA2,vA3,vA4,vA5,vA6,vA7), (out,vA1,vA2,vA3,vA4,vA5,vA6,vA7),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5); LOAD_ARG(A6); LOAD_ARG(A7);)
-#undef FW_TARGS
+    template<typename rT, class CT, typename ...aT>
+    struct return_type<rT (CT::*)(aT...) const>{
+        typedef rT type;
+        typedef CT class_type;
+        static const bool is_method = true;
+    };
 
-#define FW_TARGS class A1, class A2, class A3, class A4, class A5, class A6, class A7, class A8
-INSTANTIATE_RETURN_TYPE((A1,A2,A3,A4,A5,A6,A7,A8))
-INSTANTIATE_WRAPPERS(8, (A1,A2,A3,A4,A5,A6,A7,A8), (OSTREAM_ARG,A1,A2,A3,A4,A5,A6,A7,A8),
-                        (vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8), (out,vA1,vA2,vA3,vA4,vA5,vA6,vA7,vA8),
-                     LOAD_ARG(A1); LOAD_ARG(A2); LOAD_ARG(A3); LOAD_ARG(A4);
-                     LOAD_ARG(A5); LOAD_ARG(A6); LOAD_ARG(A7); LOAD_ARG(A8);)
-#undef FW_TARGS
+/*
+ * This is required to make GCC accept std::tuple{ (DFHack::Lua::Get<aT>(L,base++))... };
+ * The order of effects is guaranteed since C+11
+ * See: https://en.cppreference.com/w/cpp/language/eval_order (point 9)
+ */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsequence-point"
+
+#define CALL_AND_PUSH(rT,fun,args) \
+    if constexpr (std::is_same_v<rT,void>){ \
+        std::apply(fun,args); \
+    } else { \
+        rT rv = std::apply(fun,args); \
+        df::identity_traits<rT>::get()->lua_read(L, UPVAL_METHOD_NAME, &rv); \
+    }
```

Comment:
the first argument is a type, so it would have to be a templated function as types are not first class objects

Reply:
seems you're right:

```
    template<typename RT, typename FT, typename... AT>
        requires std::is_invocable_v<FT, AT...>
    void call_and_push(lua_State* L, FT fun, std::tuple<AT...>& a)
    {
        if constexpr (std::is_same_v<RT,void>){
            std::apply(fun,a);
        }
        else
        {
            RT rv = std::apply(fun, a);
            df::identity_traits<RT>::get()->lua_read(L, UPVAL_METHOD_NAME, &rv);
        }
    }

#define CALL_AND_PUSH(rT,fun,args) call_and_push<rT>(L, fun, args);
```

(requires `type_traits` header to bring in `is_invocable`; can't just use a function type here because `fun` might be a pointer to a member, which isn't a function but _is_ invocable)

Reply:
i also note that switching it to use `std::invoke` instead of `std::apply` (as above) would eliminate the explicit tuple, which should also resolve my (slight) concerns about performance

Reply:
i wasn't happy with the DRY violation in the last round, so here's this further improvement:

```
    template<typename RT, typename... AT, typename FT, typename... ET>
        requires std::is_invocable_v<FT, ET..., AT...>
    void call_and_push(lua_State* L, int base, FT fun, ET... extra)
    {
        if constexpr (std::is_same_v<RT,void>){
            std::invoke(fun, extra..., (DFHack::Lua::Get<AT>(L, base++))...);
        }
        else
        {
            RT rv = std::invoke(fun, extra..., (DFHack::Lua::Get<AT>(L, base++))...);
            df::identity_traits<RT>::get()->lua_read(L, UPVAL_METHOD_NAME, &rv);
        }
    }

    template<typename T> struct function_wrapper {};

    template<typename rT, typename ...aT>
    struct function_wrapper<rT(*)(DFHack::color_ostream&, aT...)> {
        static const int num_args = sizeof...(aT);
        static void execute(lua_State *L, int base, rT (fun)(DFHack::color_ostream& out, aT...)) {
             cur_lua_ostream_argument out(L);
             call_and_push<rT, aT...>(L, base, fun, out);
        }
    };

    template<typename rT, typename ...aT>
    struct function_wrapper<rT(*)(aT...)> {
        static const int num_args = sizeof...(aT);
        static void execute(lua_State *L, int base, rT (fun)(aT...)) {
            call_and_push<rT, aT...>(L, base, fun);
        }
    };

    template<typename rT, class CT, typename ...aT>
    struct function_wrapper<rT(CT::*)(aT...)> {
        static const int num_args = sizeof...(aT)+1;
        static void execute(lua_State *L, int base, rT(CT::*mem_fun)(aT...)) {
            CT *self = (CT*)DFHack::LuaWrapper::get_object_addr(L, base++, UPVAL_METHOD_NAME, "invoke");
            call_and_push<rT, aT...>(L, base, mem_fun, self);
        };
    };

    template<typename rT, class CT, typename ...aT>
    struct function_wrapper<rT(CT::*)(aT...) const> {
        static const int num_args = sizeof...(aT)+1;
        static void execute(lua_State *L, int base, rT(CT::*mem_fun)(aT...) const) {
            CT *self = (CT*)DFHack::LuaWrapper::get_object_addr(L, base++, UPVAL_METHOD_NAME, "invoke");
            call_and_push<rT, aT...>(L, base, mem_fun, self);
        };
    };
```

yes, there are _two_ parameter packs in the template. ain't C++ fun?

Reply:
this replacement for `call_and_push` uses `std::index_sequence` to avoid having to use postincrement and thus avoids the sequence point issue

```
    template<typename RT, typename... AT, typename FT, typename... ET, std::size_t... I>
        requires std::is_invocable_v<FT, ET..., AT...>
    void call_and_push_impl(lua_State* L, int base, std::index_sequence<I...>, FT fun, ET... extra)
    {
        if constexpr (std::is_same_v<RT, void>) {
            std::invoke(fun, extra..., (DFHack::Lua::Get<AT>(L, base+I))...);
        }
        else
        {
            RT rv = std::invoke(fun, extra..., (DFHack::Lua::Get<AT>(L, base+I))...);
            df::identity_traits<RT>::get()->lua_read(L, UPVAL_METHOD_NAME, &rv);
        }
    }

    template<typename RT, typename... AT, typename FT, typename... ET, typename indices = std::index_sequence_for<AT...> >
        requires std::is_invocable_v<FT, ET..., AT...>
    void call_and_push(lua_State* L, int base, FT fun, ET... extra)
    {
        call_and_push_impl<RT, AT...>(L, base, indices{}, fun, extra...);
    }
```

Reply:
And I was thinking that *I* was becoming a template wizard. It looks like I still have long way to go. 

> ` std::invoke(fun, extra..., (DFHack::Lua::Get<AT>(L, base+I))...);`

Three packs, two of them expanded in sync, no explicit tuple. I like this! 

Should those two functions be declared `inline`? I suppose that compilers will *probably* inline them regardless, but maybe we should nevertheless declare that.

@ab9rf Did you check whether the performance impact disappears with this version?

Reply:
I profiled the variation I have here locally and it was intermediate between baseline develop and your first proposal. I remain unconvinced that there is any performance impact; when I drill down into the numbers, there's too much variability in the timings of functions that should not be impacted by the changes for me to be convinced that there is a causal relationship and that the differences between the runs aren't just noise. A better approach to evaluating the performance impact would be to set up a bespoke harness to exercises this code specifically. Sadly, I'm not enough of a Lua programmer to come up with something to do this.

Functions declared within a class definition are always implicitly `inline` so specifying `inline` here is redundant.